modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java

```diff
@@ -326,29 +292,82 @@ ThreadPool getThreadPool() {
         return threadPool;
     }
 
-    static class Stats {
+    enum Operation {
+        GET_OBJECT("GetObject"),
+        LIST_OBJECTS("ListObjects"),
+        PUT_OBJECT("PutObject"),
+        PUT_MULTIPART_OBJECT("PutMultipartObject"),
+        DELETE_OBJECTS("DeleteObjects"),
+        ABORT_MULTIPART_OBJECT("AbortMultipartObject");
 
-        final AtomicLong listCount = new AtomicLong();
+        private final String value;
 
-        final AtomicLong getCount = new AtomicLong();
+        Operation(String value) {
+            this.value = value;
+        }
+    }
 
-        final AtomicLong putCount = new AtomicLong();
+    class StatsCollectors {
+        final ConcurrentMap<String, IgnoreNoResponseMetricsCollector> collectors;
+
+        StatsCollectors() {
+            this.collectors = Stream.of(
+                Operation.GET_OBJECT,
+                Operation.LIST_OBJECTS,
+                Operation.PUT_OBJECT,
+                Operation.PUT_MULTIPART_OBJECT,
+                Operation.DELETE_OBJECTS,
+                Operation.ABORT_MULTIPART_OBJECT
+            ).collect(Collectors.toConcurrentMap(ops -> ops.value, this::buildMetricCollector));
```

Comment:
`toConcurrentMap` puzzled me. The javadoc says `There are no guarantees on the type, mutability, or serializability of the ConcurrentMap returned.`. I think we rely on it being mutable here. The fact that a potentially non-mutable concurrent map could be returned is weird though. Regardless, I'd prefer to collect in a map and then construct a concurrent map from it.

Reply:
Thanks for the insights. You are right that the javadoc does suggest no such guarantee. I'll keep this in mind. For this PR, since I adopted your suggestion of *not* having the actual stats collection change, the concurrentMap is no longer needed. So it's now just a `Collectors.toUnmodifiableMap`.